### DISCUSSION TITLE
Optimize with variable + dataframe column

### DISCUSSION THREAD

**Discussion by jmalizzi** [kernc/backtesting.py#603] [Q&A]
Hey there friends.  I have a fully functioning ML script working with optimization on a variable called "target" but I cannot figure out how to optimize on "target" and a few dataframe columns with ML predictions.  I've spent 4 days on this so far.  :-) Any help would be appreciated.  See screenshots and code below.  Columns are RTY_Vol_pred, GC_Close_pred, S_Open_pred.

![1 screenshot after ML run](https://user-images.githubusercontent.com/100043640/155248630-68decfaf-9ba4-41df-8bc4-84cbabc259d0.jpg)
![2 screenshot after backtest](https://user-images.githubusercontent.com/100043640/155248631-76d4ea54-6c4e-45b8-b11d-0ac730a8e349.jpg)
![3 screenshot after optimization](https://user-images.githubusercontent.com/100043640/155248632-382eeda8-563b-443f-badf-2ab972ec87d0.jpg)

# a single backtest
%%time
from backtesting import Backtest, Strategy
data = df.loc[end_train:].copy()

class ML(Strategy):
        
    target = 1
    #column = RTY_Vol_pred
    
    def init(self):        
        0
        
    def next(self):       
        #if self.data.column > self.target and not self.position.is_long:
        if self.data.RTY_Vol_pred > self.target and not self.position.is_long:
            
            self.buy()
        
        for trade in self.trades:
            cur_time = self.data.index[-1]
            if cur_time - trade.entry_time >= pd.Timedelta('0 days'):
                trade.close()

bt = Backtest(data, ML, cash=1000000)
bt.run()
#bt.plot()
################################################
# optimizer
%%time

from backtesting import Backtest
import numpy as np

backtest = Backtest(data, ML, cash=1000000)

stats, heatmap = backtest.optimize(
        
    target = (.1,.2,.3,.4,.5,.6,.7,.8,.9,1.0,1.1,1.2,1.3,1.4,1.5),
    #column = 
    
    #constraint=lambda p: p.n_exit < p.n_enter < p.n1 < p.n2,
    #maximize='Equity Final [$]',
    maximize='Sharpe Ratio',
    max_tries=200,
    random_state=0,
    return_heatmap=True)

print(heatmap)

**Comment by kernc** [comment]
_What_ exactly would you like to optimize? We optimize [free parameters](https://en.wikipedia.org/wiki/Free_parameter) of the model (logic), not fixed values of `target` and `RTY_Vol_pred` which you pass in as `data` ... :thinking:

**Reply by jmalizzi** [reply]
@kernc  i want to optimize through 200+ columns of ML regression predictions i.e. RTY_Vol_pred and a range of target variables which represent the value of the predictions.  the 200+ columns are being created before the backtest begins.  my example above only shows 3 of the columns as an example.
